### DISCUSSION TITLE
关于“Async/Await 语法糖语义不明”的修复方案

### DISCUSSION THREAD

**Discussion by XiaoHeitu** [dotnet/csharplang#6570] [Language Ideas]
先来一段简单的代码：
![image](https://user-images.githubusercontent.com/38452227/196588263-40ddf421-92f9-4b31-8db8-e51d7ee4f1a1.png)

这段代码相信大家都能看懂。代码所代表的含义吧。。但是就是这样含义准确的代码。在现在语法下得不到正确的执行顺序。
官方是不是应该反思一下。什么叫着 Async 。为什么标记为 Async的方法并不按 Async 的方式去执行？

**Comment by HaloFour** [comment]
The `async` keyword does not start a new thread as that would have considerable overhead.  What it does is allow the method to be rewritten by the compiler to be a coroutine, where `await` expressions cause the execution of the method to be "suspended" until the awaited expression completes.

**Reply by XiaoHeitu** [reply]
为什么不反过来实现呢。当未使用 await 时，重写为协程，当使用await 时当成 同步方法使用。这样开销不是更少？

**Reply by theunrepentantgeek** [reply]
Google translation:

> Why not do it the other way around. When await is not used, it is rewritten as a coroutine, and when await is used, it is used as a synchronous method. This way the overhead is not less？

The answer to any question about "Why not" is that no-one has invested the time and energy to design, implement and test the feature.

(Aside: Most readers won't bother to use machine translation to convert your words to English. You'll get better engagement with your ideas if you translate them yourself.)

**Reply by HaloFour** [reply]
The `async` keyword is necessary in order to inform the compiler that `await` is always a keyword within the method, otherwise it could be parsed ambiguously.  The team decided that it's probably also a lot better to explicitly mark the method since the coroutine machinery changes how it behaves considerably.

I believe that if the team could reverse time they probably would have done the same with iterators, require an `iterator` keyword or the like.   That's why two keywords are required, `yield return` or `yield break`, because `yield` by itself would have the same ambiguous parsing problems.

**Reply by XiaoHeitu** [reply]
> Google translation:
> 
> > Why not do it the other way around. When await is not used, it is rewritten as a coroutine, and when await is used, it is used as a synchronous method. This way the overhead is not less？
> 
> The answer to any question about "Why not" is that no-one has invested the time and energy to design, implement and test the feature.
> 
> (Aside: Most readers won't bother to use machine translation to convert your words to English. You'll get better engagement with your ideas if you translate them yourself.)

谢谢你的回答，你给出了一个最接近真实的答案。同也感谢你关于翻译的好意提醒。但我觉得现在应该也有不少人在使用翻译工具，将英文翻译成汉语。所以我这样的方式来参与讨论，可以更好的让他们了解我的想法和思路。

**Comment by XiaoHeitu** [comment]
我的想法是。。。Async 应该要能独立使用。不能强制将 Aaync 与 Await 绑定使用。就像上面的代码一样。只要标记为 Async 就在没有使用Await 时，就一定要是异步执行 才对。不能因为 方法内部没有 Await ，而否定了 方法上面的 Async 标记。

**Reply by theunrepentantgeek** [reply]
Google translation:

> My thoughts are... Async should be able to code independently. You cannot force the use of Aaync with Await. Just like the code above. As long as it is marked as Async and when Await is not used, it must be asynchronous execution. The Async tag on the method cannot be negated because there is no Await inside the method.

The **async** keyword indicates only that the compiler should rewrite the code to allow suspension when an **await** is used. 
It is nothing to do with threading, and everything to do with the compiler constructing a state machine to run the method incrementally.

If you want the code to run in a separate thread, there are easy ways to do that.

**Reply by HaloFour** [reply]
This feature has already shipped with the current behavior and will not be changed.  It's very intentional that it does not start a new thread.  As mentioned, that would add considerable overhead.  Creating threads is almost always the wrong thing to do, especially as I/O bound operations typically don't require any threads.  If you want to execute a task on a separate thread you can use [`Task.Run`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.run?view=net-7.0).

**Reply by XiaoHeitu** [reply]
> This feature has already shipped with the current behavior and will not be changed. It's very intentional that it does not start a new thread. As mentioned, that would add considerable overhead. Creating threads is almost always the wrong thing to do, especially as I/O bound operations typically don't require any threads. If you want to execute a task on a separate thread you can use [`Task.Run`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.run?view=net-7.0).

虽然不能改变什么。但我们想提出疑问。。。就像你说的一样，如果在现有的Async/Await的语法要求下能不开始新线程做到Main与FooA并行。那上面的代码也能做到不启动新线程的吧。

**Reply by HaloFour** [reply]
Two CPU-bound loops cannot execute concurrently without starting a new thread.  That's the situation you run into with the code you posted, the loop in `FooA` runs forever and control is never returned to `Main`.

**Reply by XiaoHeitu** [reply]
> Two CPU-bound loops cannot execute concurrently without starting a new thread. That's the situation you run into with the code you posted, the loop in `FooA` runs forever and control is never returned to `Main`.

好吧。。这可能并不是我的本意。我将代码进行一次修改。。

**Reply by qrli** [reply]
Async is a broad class of behaviors - it may be async from start, async from middle, async conditionally, or not async at all but with async method contract. The correct choice depends.

异步并非只有一种模式，而是广泛的一类模式。它可以从一开始执行就异步，也可能从中间才开始异步，也可能在特定条件下异步，甚至完全同步，即使方法的声明是异步。要视实际业务需求而定。

Marking a method `async` mean you want to bring in the async cost and basic behavior difference, but it does not mean the method has to do async. There are practical cases it will run synchronously to the end. 

将方法标记为async只是表示你要引入异步方法的行为，以及它的性能和代码体积的代价，但并不意味着这个方法必须以异步方式工作。有些情况下，同步执行到底也会是个有意义的选择。

Regarding the behavior difference: without `async`, any exception from `FooA()` will pop up from `_ = FooA()`; however, with `async`, you will not see the exception as you discarded the result task. 

关于异步方法的行为差别：在没有async时，FooA()中抛的异常会在`_ = FooA()`这行抛出，而有async时，你将不会收到异常，因为异常被包含在返回的task里，而你的代码丢弃了这个task。

**Comment by AartBluestoke** [comment]
the while loop in main can only progress once FooA has returned something. async doesn't change that.
However
in an async function 'await' allows an early return of a potentially incomplete task from the inner layer.

this inner task is typically created when there is an IO task - "i have asked for some data from disk, i am returning the CPU back to you for a while".

It is assumed that somewhere something will 'await' that task to allow it to be continued. Discarding that Task can cause problems.

**Reply by XiaoHeitu** [reply]
> the while loop in main can only progress once FooA has returned something. async doesn't change that. However in an async function 'await' allows an early return of a potentially incomplete task from the inner layer.
> 
> this inner task is typically created when there is an IO task - "i have asked for some data from disk, i am returning the CPU back to you for a while".
> 
> It is assumed that somewhere something will 'await' that task to allow it to be continued. Discarding that Task can cause problems.

我是认为，只要FooA方法标记为 Async ，并且在调用是未使用Await关键字时。开发人员就已经默认了应该方法返回值不会对系统造成影响可以丢弃。所以也就不会出现你所说的问题。如果需要得到FooA 的返回值时才会去使用 Await 关键字。

**Comment by huoyaoyuan** [comment]
不要混淆`Task.Run`和`async Task F()`。你的代码很清晰：先执行Foo里的所有东西，再返回`Task.CompletedTask`，然后再执行`Main`剩下的不放呢。

Don't get confused with `Task.Run` and `async Task F()`. Your code is very clear that execute everything in `Foo`, then get `Task.CompletedTask` returned, then executing the remaining of `Main`.
